In [1]:
import pandas as pd
import numpy as np

# Import data

In [3]:
data = pd.read_csv('Downloads/transaction_data.csv')

In [4]:
data.head()

txn_time    txn_date   order_id        uid  shop_id  \
0  10/30/2019 13:37  10/30/2019  142008725  102161515     1005   
1  10/30/2019 20:26  10/30/2019  142119377  102867263    10050   
2  10/30/2019 12:22  10/30/2019  141986418  100401383    10075   
3   10/30/2019 9:46  10/30/2019  141941958  103213155    10088   
4  10/30/2019 21:56  10/30/2019  142139275  101003290    10142   

   shop_owner_uid     gmv  rebate  
0       100054430   70000   20000  
1         1792841   80000   20000  
2       100155720  125000   20000  
3       103894677   82000   20000  
4       103909421  260000   20000

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4611 entries, 0 to 4610
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   txn_time        4611 non-null   object
 1   txn_date        4611 non-null   object
 2   order_id        4611 non-null   int64 
 3   uid             4611 non-null   int64 
 4   shop_id         4611 non-null   int64 
 5   shop_owner_uid  4611 non-null   int64 
 6   gmv             4611 non-null   int64 
 7   rebate          4611 non-null   int64 
dtypes: int64(6), object(2)
memory usage: 288.3+ KB


# Cleaning

In [7]:
data['txn_date'] = pd.to_datetime(data['txn_date'])
data['txn_time'] = pd.to_datetime(data['txn_time'])

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4611 entries, 0 to 4610
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   txn_time        4611 non-null   datetime64[ns]
 1   txn_date        4611 non-null   datetime64[ns]
 2   order_id        4611 non-null   int64         
 3   uid             4611 non-null   int64         
 4   shop_id         4611 non-null   int64         
 5   shop_owner_uid  4611 non-null   int64         
 6   gmv             4611 non-null   int64         
 7   rebate          4611 non-null   int64         
dtypes: datetime64[ns](2), int64(6)
memory usage: 288.3 KB


In [9]:
data.head()

txn_time   txn_date   order_id        uid  shop_id  \
0 2019-10-30 13:37:00 2019-10-30  142008725  102161515     1005   
1 2019-10-30 20:26:00 2019-10-30  142119377  102867263    10050   
2 2019-10-30 12:22:00 2019-10-30  141986418  100401383    10075   
3 2019-10-30 09:46:00 2019-10-30  141941958  103213155    10088   
4 2019-10-30 21:56:00 2019-10-30  142139275  101003290    10142   

   shop_owner_uid     gmv  rebate  
0       100054430   70000   20000  
1         1792841   80000   20000  
2       100155720  125000   20000  
3       103894677   82000   20000  
4       103909421  260000   20000

In [14]:
data.isnull().sum()

txn_time          0
txn_date          0
order_id          0
uid               0
shop_id           0
shop_owner_uid    0
gmv               0
rebate            0
dtype: int64

# EDA

In [17]:
data['gmv'].mean()

138985.27586206896

In [22]:
conditions = [
    (data['gmv'] <= 100000),
    (data['gmv'] > 100000) & (data['gmv'] <= 200000),
    (data['gmv'] > 200000) & (data['gmv'] <= 300000),
    (data['gmv'] > 300000) & (data['gmv'] <= 400000),
    (data['gmv'] > 400000) & (data['gmv'] <= 500000),
    (data['gmv'] > 500000) & (data['gmv'] <= 600000),
    (data['gmv'] > 600000) & (data['gmv'] <= 700000),
    (data['gmv'] > 700000) & (data['gmv'] <= 800000),
    (data['gmv'] > 800000) & (data['gmv'] <= 900000),
    (data['gmv'] > 900000) & (data['gmv'] <= 1000000),
    (data['gmv'] > 1000000)
    ]

In [23]:
values = ['0-1', '1-2', '2-3', '3-4', '4-5', '5-6', '6-7', '7-8', '8-9', '9-10', '>1tr']

In [24]:
data['Group'] = np.select(conditions, values)

In [25]:
data.head()

txn_time   txn_date   order_id        uid  shop_id  \
0 2019-10-30 13:37:00 2019-10-30  142008725  102161515     1005   
1 2019-10-30 20:26:00 2019-10-30  142119377  102867263    10050   
2 2019-10-30 12:22:00 2019-10-30  141986418  100401383    10075   
3 2019-10-30 09:46:00 2019-10-30  141941958  103213155    10088   
4 2019-10-30 21:56:00 2019-10-30  142139275  101003290    10142   

   shop_owner_uid     gmv  rebate    tier Group  
0       100054430   70000   20000  tier_4   0-1  
1         1792841   80000   20000  tier_4   0-1  
2       100155720  125000   20000       0   1-2  
3       103894677   82000   20000  tier_4   0-1  
4       103909421  260000   20000       0   2-3

In [28]:
data.groupby(['Group'])['order_id'].count()

Group
0-1     2994
1-2      897
2-3      312
3-4      139
4-5       94
5-6       57
6-7       26
7-8       26
8-9       21
9-10       9
>1tr      36
Name: order_id, dtype: int64

Các đơn hàng có giá trị từ 100k trở xuống là khoảng có số lượng giao dịch nhiều nhất với 2994 giao dịch

In [33]:
data["ID"] = data["uid"].astype(str) + '-' + data["shop_id"].astype(str)

In [34]:
data.head()

txn_time   txn_date   order_id        uid  shop_id  \
0 2019-10-30 13:37:00 2019-10-30  142008725  102161515     1005   
1 2019-10-30 20:26:00 2019-10-30  142119377  102867263    10050   
2 2019-10-30 12:22:00 2019-10-30  141986418  100401383    10075   
3 2019-10-30 09:46:00 2019-10-30  141941958  103213155    10088   
4 2019-10-30 21:56:00 2019-10-30  142139275  101003290    10142   

   shop_owner_uid     gmv  rebate    tier Group               ID  
0       100054430   70000   20000  tier_4   0-1   102161515-1005  
1         1792841   80000   20000  tier_4   0-1  102867263-10050  
2       100155720  125000   20000       0   1-2  100401383-10075  
3       103894677   82000   20000  tier_4   0-1  103213155-10088  
4       103909421  260000   20000       0   2-3  101003290-10142

In [57]:
data_copy = data.copy()

In [58]:
data_copy.drop(data_copy[data_copy['rebate'] < 1].index, inplace = True)

In [59]:
dup = data_copy[data_copy.duplicated(['ID'], keep=False)].sort_values(by = ['ID'])

In [64]:
dup.head()

txn_time   txn_date   order_id        uid  shop_id  \
1116 2019-10-31 13:02:00 2019-10-31  142246468  100024287     1503   
3330 2019-11-02 12:38:00 2019-11-02  142734855  100024287     1503   
2619 2019-11-01 17:42:00 2019-11-01  142574513  100043020     5577   
3896 2019-11-02 15:33:00 2019-11-02  142781339  100043020     5577   
6    2019-10-30 12:46:00 2019-10-30  141993339  100060089    10173   

      shop_owner_uid     gmv  rebate    tier Group               ID  
1116         1744747  205000   20000       0   2-3   100024287-1503  
3330         1744747  120000   20000       0   1-2   100024287-1503  
2619       103431327  100000   20000  tier_4   0-1   100043020-5577  
3896       103431327  100000   20000  tier_4   0-1   100043020-5577  
6          103915476   56000   16800  tier_4   0-1  100060089-10173

In [62]:
dup.min()

txn_time          2019-10-30 00:23:00
txn_date          2019-10-30 00:00:00
order_id                    141905033
uid                           1130584
shop_id                           136
shop_owner_uid                1005510
gmv                             10000
rebate                           3000
tier                                0
Group                             0-1
ID                     100024287-1503
dtype: object

In [66]:
oIDs = dup.groupby('shop_owner_uid')['shop_owner_uid'].count()

In [70]:
oIDs.sort_values()

shop_owner_uid
1005510        2
102928117      2
102946203      2
102949530      2
102986571      2
            ... 
103280547     14
100305004     16
103333474     16
1952823       18
102817559    132
Name: shop_owner_uid, Length: 252, dtype: int64